pkg/Cpanel/Security/Advisor/Assessors/Usernames.pm

```diff
@@ -0,0 +1,178 @@
+package Cpanel::Security::Advisor::Assessors::Usernames;
+
+# Copyright (c) 2018, cPanel, Inc.
+# All rights reserved.
+# http://cpanel.net
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#     * Redistributions of source code must retain the above copyright
+#       notice, this list of conditions and the following disclaimer.
+#     * Redistributions in binary form must reproduce the above copyright
+#       notice, this list of conditions and the following disclaimer in the
+#       documentation and/or other materials provided with the distribution.
+#     * Neither the name of the owner nor the names of its contributors may
+#       be used to endorse or promote products derived from this software
+#       without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL  BE LIABLE FOR ANY
+# DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+# ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+# SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use warnings;
+use strict;
+
+use Cpanel::Config::LoadUserOwners   ();
+use Cpanel::Config::LoadUserDomains  ();
+use Cpanel::ArrayFunc::Uniq          ();
+use Cpanel::Validate::Username::Core ();
+
+use base 'Cpanel::Security::Advisor::Assessors';
+
+my $GO_URL = "https://go.cpanel.net/usernames";
+
+my $update_user_domains_cmd = '/scripts/updateuserdomains --force';
+my $rename_user_cmd         = 'whmapi1 modifyacct user=old_username newuser=new_username';
+
+sub generate_advice {
+    my ($self) = @_;
+    $self->_check_for_invalid_users;
+
+    return 1;
+}
+
+sub _check_for_invalid_users {
+    my ($self) = @_;
+
+    my @users;
+
+    # gather the users in /etc/trueuserowners file.
+    my %trueuserowners;
+    Cpanel::Config::LoadUserOwners::loadtrueuserowners( \%trueuserowners, 1, 1 );
+    my @true_users = ( keys %trueuserowners );
+
+    # gather users that exist in /var/cpanel/userdata directory
+    my $user_data_dir = '/var/cpanel/userdata';
+    opendir( my $userdata_dh, $user_data_dir );
+    my @userdata_users = grep { $_ !~ /\A\.{1,2}\z|\Anobody\z/ && -e "$user_data_dir/$_/main" } readdir $userdata_dh;
```

Comment:
In Perl, it's idiomatic to avoid writing `$_` explicitly like this.  Writing the following would be more common:

`grep { !/\A\.{1,2}\z|\Anobody\z/ && -e "$user_data_dir/$_/main" }`